packages/lodestar/src/chain/validation/block.ts

```diff
@@ -26,56 +40,85 @@ export async function validateGossipBlock(
     });
   }
 
-  const currentSlotWithGossipDisparity = chain.clock.currentSlotWithGossipDisparity;
-  if (currentSlotWithGossipDisparity < blockSlot) {
-    throw new BlockGossipError(GossipAction.IGNORE, {
-      code: BlockErrorCode.FUTURE_SLOT,
-      currentSlot: currentSlotWithGossipDisparity,
-      blockSlot,
-    });
+  // Check if the block is already known. We know it is post-finalization, so it is sufficient to check the fork choice.
+  //
+  // In normal operation this isn't necessary, however it is useful immediately after a
+  // reboot if the `observed_block_producers` cache is empty. In that case, without this
+  // check, we will load the parent and state from disk only to find out later that we
+  // already know this block.
+  const blockRoot = config.getForkTypes(block.slot).BeaconBlock.hashTreeRoot(block);
+  if (chain.forkChoice.getBlock(blockRoot) !== null) {
+    throw new BlockGossipError(GossipAction.IGNORE, {code: BlockErrorCode.ALREADY_KNOWN, root: blockRoot});
   }
 
   // No need to check for badBlock
   // Gossip de-duplicates messages so we shouldn't be able to receive a bad block twice
 
-  const existingBlock = await db.block.get(blockRoot);
-  if (existingBlock?.message.proposerIndex === block.message.proposerIndex) {
+  // [IGNORE] The block is the first block with valid signature received for the proposer for the slot, signed_beacon_block.message.slot.
+  const proposerIndex = block.proposerIndex;
+  if (chain.seenBlockProposers.isKnown(blockSlot, proposerIndex)) {
+    throw new BlockGossipError(GossipAction.IGNORE, {code: BlockErrorCode.REPEAT_PROPOSAL, proposerIndex});
+  }
+
+  // [REJECT] The current finalized_checkpoint is an ancestor of block -- i.e.
+  // get_ancestor(store, block.parent_root, compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)) == store.finalized_checkpoint.root
+  const parentRoot = block.parentRoot.valueOf() as Uint8Array;
+  const parentBlock = chain.forkChoice.getBlock(parentRoot);
+  if (parentBlock === null) {
+    // If fork choice does *not* consider the parent to be a descendant of the finalized block,
+    // then there are two more cases:
+    //
+    // 1. We have the parent stored in our database. Because fork-choice has confirmed the
+    //    parent is *not* in our post-finalization DAG, all other blocks must be either
+    //    pre-finalization or conflicting with finalization.
+    // 2. The parent is unknown to us, we probably want to download it since it might actually
+    //    descend from the finalized root.
+    if ((await db.block.get(parentRoot as Uint8Array)) !== null) {
```

Comment:
this never happen with the current implementation. When `archiveBlocks` runs, if it found any blocks conflicting to the finalized block it'll delete it from db

Just afraid we query db for nothing here.

Reply:
That's a good point, and this check won't save us from blocks from unknown forks if they are two children deep from our latest known forked block. I'll remove